### PR TITLE
Surface free-write corrections as inline highlights and a corrections list

### LINKED ISSUE #312: Free-write feedback never shows where the errors are: LLM-generated corrections are discarded before reaching the user
## Current behavior

When a `free_write` exercise is evaluated, the user gets prose feedback, an explanation and a score — but no indication of **where** the errors are. Example from a real session (German, A2): the answer *"Meine letzte Reise war nach Deutschalnd . Ich habe viele abenteuer gehabt. Ich bin mit auto gefahren."* scores 80%, the feedback opens with "Great job!", the textarea border turns red — and the user has no way to tell which words cost them the 20% (the misspelled *Deutschalnd*, lowercase *abenteuer*, the missing article in *mit auto*).

The ingredients for a much clearer display already exist but are thrown away:

- The evaluation prompt (`FREE_WRITE_EVAL_PROMPT` in `backend/app/services/prompts/lesson.py`) already asks the LLM for `corrections: [{"original", "corrected", "explanation"}]`, and `FreeWriteEvaluation` (`backend/app/schemas/lessons.py:119-122`) parses them.
- The router then discards them: `backend/app/routers/lessons.py:389-396` reads only `score` and `feedback`. `ExerciseAnswerResponse` has no `corrections` field and the `Exercise` model has no column for them, so the data never reaches the frontend.
- The frontend (`frontend/src/app/(app)/lesson/[id]/page.tsx`) renders feedback and explanation as plain paragraphs.

A related signal mismatch on the same surface: `isAnswerCorrect = (score ?? 0) >= 1` (`page.tsx:497-498`), so a partially-correct answer (e.g. 0.8) gets the same red border and ✕ badge as a completely wrong one, while the feedback text is praising the answer.

## Proposed UX

```
┌─ YOUR ANSWER (read-only after evaluation) ───────────┐
│ Meine letzte Reise war nach ~~Deutschalnd~~          │
│ **Deutschland**. Ich habe viele ~~abenteuer~~        │
│ **Abenteuer** gehabt. Ich bin ~~mit auto~~           │
│ **mit dem Auto** gefahren.                           │
│                                                      │
│   ~~struck through~~ = red   (the error)             │
│   **bold**           = green (correction, inline)    │
└──────────────────────────────────────────────────────┘

┌─ CORRECTIONS (3) ────────────────────────────────────┐
│ ✗ Deutschalnd  →  ✓ Deutschland                      │
│   Spelling of the country name.                      │
│ ✗ abenteuer    →  ✓ Abenteuer                        │
│   Nouns are capitalized in German.                   │
│ ✗ mit auto     →  ✓ mit dem Auto                     │
│   Dative article after 'mit'.                        │
└──────────────────────────────────────────────────────┘

SCORE 80%   ← amber border (partially correct), not red
```

- After evaluation the submitted answer is rendered read-only with each matched `original` span struck through (red), immediately followed by the `corrected` text (green). Not color-only: strikethrough marks the error and the correction is bold/underlined, so the distinction survives color-blindness (WCAG 1.4.1).
- Below the feedback block, an always-visible **Corrections** list: `original → corrected` plus the per-error explanation (already produced in the user's native language). This list is also the accessibility/mobile path — no hover-only tooltips.
- Score styling gains a partial state: green for `score >= 1`, amber when corrections exist and the score is partial, red otherwise (replaces today's binary green/red border at `page.tsx:954-957`).

## Technical sketch

**Backend** (the evaluation prompt text itself needs no change):

- Type the corrections — `corrections: list[FreeWriteCorrection]` with `original/corrected/explanation: str` — instead of `list[dict]`. Note this tightens the JSON schema passed to `structured_output()` for a prompt listed as behavior-critical in `specs/prompts.instructions.md`; it is a deliberate change, and the existing structured-output parsing tests must be kept/extended per that spec's maintenance rules.
- Persist corrections alongside `feedback` (new nullable JSON column on `Exercise`) so they survive a page reload, and expose them in `ExerciseAnswerResponse` / `ExerciseResponse`.

**Frontend**:

- Highlight spans are computed client-side by locating each `original` substring in the submitted answer (whitespace/case-normalized matching). No character offsets are requested from the LLM — LLM-produced offsets are unreliable.
- Degradation paths: (a) a correction whose `original` cannot be matched in the answer is shown in the list only; (b) empty `corrections` with `score < 1` falls back to today's prose-only display; (c) overlapping matches keep the first and push the rest to the list.
- New i18n label for the corrections section in all `messages/*.json` locales.

Scope: only the free-write answer path (`POST /exercises/{exercise_id}/answer` + the lesson detail page). Lesson generation is untouched.

### PR DESCRIPTION
**Pull request by arqo123**
Closes #312

## What

Free-write feedback never showed **where** the errors are: the evaluation prompt already returns per-error correction objects (`original`, `corrected`, `explanation`), but the answer endpoint read only `score` and `feedback` and dropped them. This PR carries the corrections through to the UI:

- **Inline annotations** — after evaluation the submitted answer is rendered read-only with each matched `original` fragment struck through (red) and the `corrected` text right after it (green, underlined). Not color-only: strikethrough + underline carry the signal for color-blind users (WCAG 1.4.1).
- **Corrections list** — an always-visible block under the feedback with each `original → corrected` pair and its native-language explanation. This is also the accessibility/mobile path (no hover-only tooltips) and the fallback for corrections whose fragment can't be located in the answer.
- **Amber partial state** — a free-write answer with corrections and `0 < score < 1` now gets an amber border and `±` badge instead of the old binary green/red (an 80% answer no longer shows a red frame next to praising feedback). As proposed in the issue, amber requires corrections to be present, so the LLM-unavailable fallback (score 0.5, no corrections) keeps the red state it had before.

### Screenshots (local run, seeded data)

Annotated answer + corrections list (score 0.8, three corrections):

Annotated free-write answer with inline corrections and a corrections list

Degradation path — same score, LLM returned no corrections → today's prose-only display with the amber state:

Fallback prose-only feedback with amber partial state

## How

**Backend**

- `FreeWriteCorrection` model (`original` and `corrected` required, `explanation` defaulting to `""`); `FreeWriteEvaluation.corrections` is now `list[FreeWriteCorrection]` instead of `list[dict]`, with a before-validator that drops entries lacking a nonblank `original`/`corrected` pair. `structured_output()` parses client-side (JSON mode), so this validator is where malformed correction objects are caught — they are discarded instead of failing the whole evaluation. The `FREE_WRITE_EVAL_PROMPT` text is untouched.
- Corrections are persisted on the exercise (`exercises.corrections`, nullable JSON; Alembic revision `0052_exercise_corrections`) and exposed in `ExerciseAnswerResponse` and the lesson-detail `ExerciseResponse`, so they survive a reload and review mode. The LLM-fallback path leaves them `NULL`.

**Frontend**

- `annotateAnswer()` (`frontend/src/lib/free-write-corrections.ts`) locates each `original` fragment in the submitted answer client-side (exact, then case-insensitive, then trimmed match — no LLM-produced offsets) and splits it into plain/fix segments. Each correction gets its own non-overlapping occurrence: repeated fragments take successive occurrences, whole-word matches are preferred over matches inside a longer word, and the longer fragment wins when two start at the same position. Corrections that cannot be placed yield no inline segment and appear only in the list; empty `corrections` falls back to the existing prose-only display.
- Lesson page renders the annotated block in place of the disabled textarea when at least one fragment matched, plus the corrections list; new `lesson.corrections` i18n key added to all 10 locales.

## Tests

- Backend: new `test_answer_free_write_returns_and_persists_corrections` (answer response + lesson-detail persistence), corrections assertions added to the existing free-write answer/fallback tests. Schema and router tests for dropped malformed corrections. Full suite: 1023 passed, coverage 85.58%.
- Frontend: `frontend/tests/lib/free-write-corrections.test.ts` covers exact/case-insensitive/trimmed matching, unmatched corrections, overlaps, repeated fragments, whole-word vs. subword collisions, and empty originals. `frontend/tests/app/lesson-free-write-corrections.test.tsx` renders the lesson page end to end: submit → inline `<del>`/`<ins>` markup, corrections list (including an unmatched correction), amber `±` state; review mode with persisted corrections; the 0.5 fallback staying red. `npm run lint`, `tsc --noEmit`, and `vitest run` (496 tests) all pass.
- Migration verified against a scratch database (column added, chain `0051_conversation_speech_pause → 0052_exercise_corrections`).

Specs updated: `api-endpoints`, `database-models`, `study-plan`, `services`, `prompts`. `CHANGELOG.md` and `specs/version.md` are left untouched — release bookkeeping for the maintainer.

### CODE REVIEW

File: frontend/src/lib/free-write-corrections.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+// still shown in the corrections list below the answer.
+export function annotateAnswer(
+  answer: string,
+  corrections: FreeWriteCorrection[]
+): AnswerSegment[] {
+  const matches: Array<{ start: number; end: number; corrected: string }> = []
+  for (const correction of corrections) {
+    if (!correction.original || !correction.corrected) continue
+    const match = findMatch(answer, correction.original)
+    if (match) matches.push({ ...match, corrected: correction.corrected })
```
Comment (artcc) [inline frontend/src/lib/free-write-corrections.ts]: `findMatch()` always searches from the beginning of the answer, so this can annotate the wrong occurrence or drop repeated corrections. For example, correcting `in` in `Ich bin in Berlin` matches the substring inside `bin`; two corrections for two identical fragments both resolve to the first occurrence and the second is later discarded as overlapping. Please allocate non-overlapping occurrences (and prefer the most specific match when candidates start at the same position) and add tests for repeated fragments and word/subword collisions.

File: backend/app/schemas/lessons.py [hunk 2]
```diff
@@ -109,17 +110,24 @@ class ExerciseAnswerRequest(BaseModel):
     answer: str
 
 
+class FreeWriteCorrection(BaseModel):
+    original: str = ""
+    corrected: str = ""
+    explanation: str = ""
```
Comment (artcc) [inline backend/app/schemas/lessons.py:116]: These defaults make malformed LLM objects such as `{}` or `{"original": "foo"}` valid structured output. The router then persists them, and the UI can render an incomplete `foo ->` correction row. This is not equivalent to the old behavior now that corrections are user-visible. Please require nonblank `original` and `corrected` values, or sanitize unusable entries before persistence; tolerating a missing explanation is reasonable.

File: frontend/src/app/(app)/lesson/[id]/page.tsx [hunk 3]
```diff
@@ -497,6 +503,22 @@ export default function LessonPage() {
   const exercise = exercises[currentExercise]
   const isEvaluated = exercise?.score !== null
   const isAnswerCorrect = (exercise?.score ?? 0) >= 1
+  const isPartiallyCorrect =
+    exercise?.exercise_type === 'free_write' &&
+    isEvaluated &&
+    !isAnswerCorrect &&
+    (exercise?.score ?? 0) > 0
```
Comment (artcc) [inline frontend/src/app/(app)/lesson/[id]/page.tsx]: Please confirm or adjust this semantic change. The backend uses `score = 0.5` when free-write evaluation is unavailable, with `corrections = null`, so this condition labels an unevaluated answer as partially correct (`±`/amber). The issue proposed amber only when corrections exist and the score is partial. A neutral/unavailable state may be more accurate for the fallback; whichever behavior is intended should have a test.

File: frontend/src/app/(app)/lesson/[id]/page.tsx [hunk 4]
```diff
@@ -323,6 +328,7 @@ export default function LessonPage() {
           ...copy[currentExercise],
           score: result.score,
           feedback: result.feedback,
+          corrections: result.corrections ?? null,
```
Comment (artcc) [inline frontend/src/app/(app)/lesson/[id]/page.tsx:331]: The utility tests cover segmentation, but the main user-facing flow is currently untested. Please add a focused lesson-page test proving that response corrections enter state and render the inline `<del>/<ins>` markup, the corrections list, and the amber partial state. Ideally it should also cover unmatched corrections remaining visible in the list or persisted corrections loaded in review mode.

**Review by artcc — CHANGES_REQUESTED** [review]
The overall backend/API/migration flow is sound: corrections are persisted and survive reload/review, the API change is additive, the Alembic chain is valid, and CI passes. I found a few cases to address before merge:

1. Inline matching can select the wrong substring and loses repeated identical corrections.
2. Empty defaults allow incomplete correction objects to be accepted, persisted, and rendered.
3. The principal lesson-page rendering/state flow needs a focused integration test.
4. The amber state currently also represents the LLM-unavailable `0.5` fallback as partially correct; please confirm or adjust that behavior and test it.

I will handle the repository documentation/release bookkeeping separately as maintainer, including changelog, version, affected architecture specs, and test-count synchronization.

File: frontend/src/lib/free-write-corrections.ts (hunk 1 above)
Comment (arqo123) [reply frontend/src/lib/free-write-corrections.ts]: Good catch — the single `indexOf` was behind both the subword hit and the lost repeated corrections. `annotateAnswer()` now collects every occurrence of each fragment and allocates them greedily in answer order, one span per correction, so two identical fragments land on successive occurrences instead of both resolving to the first. Whole-word occurrences (Unicode-aware `\p{L}\p{N}` boundary check) are allocated in a first pass, so `in` in `Ich bin in Berlin` picks the standalone word and only falls back to a subword match when no whole-word occurrence is free. When two candidates start at the same position the longer fragment wins.

Tests added for repeated fragments, the `bin`/`in` collision, the subword fallback, same-start preference, and a shorter fragment being pushed to its next occurrence when a longer one already covers the first.

File: backend/app/schemas/lessons.py (hunk 2 above)
Comment (arqo123) [reply backend/app/schemas/lessons.py:116]: Agreed, that was a leftover from matching the old `list[dict]` tolerance. `original` and `corrected` are now required on `FreeWriteCorrection` (`explanation` still defaults to `""`), and `FreeWriteEvaluation` has a `mode="before"` validator that drops entries without a nonblank pair. Since `structured_output()` runs in plain JSON mode this validator is the only place the shape gets checked, and dropping the bad entry keeps the rest of the evaluation usable instead of failing the whole response into the 0.5 fallback. The frontend list filter also went from `original || corrected` to `&&`, which is what let the `foo ->` row through.

Tests: schema-level (`{}`, `{"original": "foo"}`, blank strings and non-object entries dropped; a valid entry without explanation kept; direct construction without a field raises) plus a router test asserting only the usable entry is returned and persisted.

File: frontend/src/app/(app)/lesson/[id]/page.tsx (hunk 3 above)
Comment (arqo123) [reply frontend/src/app/(app)/lesson/[id]/page.tsx]: You're right, this labelled an unevaluated answer as partially correct. I went with the rule from the issue: amber only when corrections are present and `0 < score < 1`. The fallback (0.5, `corrections = null`) therefore keeps the red ✕ it had before this PR.

I did consider a dedicated neutral state, but the API has no signal for "evaluation unavailable" today — inferring it from `score === 0.5 && corrections === null` would misfire on a genuine 0.5 evaluation — so that needs a persisted flag on the exercise and feels like a separate issue. Both cases are covered in the new lesson-page test, and I've corrected the PR description, which argued for score-only keying.

File: frontend/src/app/(app)/lesson/[id]/page.tsx (hunk 4 above)
Comment (arqo123) [reply frontend/src/app/(app)/lesson/[id]/page.tsx:331]: Added `frontend/tests/app/lesson-free-write-corrections.test.tsx`. It renders the real lesson page with `apiFetch` mocked and covers:

- submitting a free-write answer: the response corrections enter state and render the `<del>`/`<ins>` annotations, the amber border and `±` badge, and the corrections list — including a correction whose fragment is not in the answer, which appears only in the list;
- review mode (`is_completed: true`): persisted corrections from the lesson detail payload render the same way;
- the 0.5/`null` fallback keeping the red state with no list;
- a partial score with an empty corrections array staying red.

**Comment by arqo123** [comment]
Thanks for the review — all four points addressed in 8e509d2 (rebased on the current branch head, which includes the develop merge):

1. `annotateAnswer()` allocates one non-overlapping occurrence per correction, prefers whole-word matches, and takes the longer fragment when two start at the same position.
2. `FreeWriteCorrection` requires nonblank `original`/`corrected`; malformed entries are dropped in `FreeWriteEvaluation` before persistence.
3. Amber now requires corrections to be present (the rule from the issue); the 0.5 fallback stays red. PR description updated accordingly.
4. New lesson-page integration test for submit, review mode, and the fallback.

`specs/prompts.instructions.md` updated to match the stricter correction schema. Backend: 1023 passed, coverage 85.58%. Frontend: lint, `tsc --noEmit`, and 496 tests pass. Changelog/version left to you as before.

File: frontend/src/lib/free-write-corrections.ts (hunk 1 above)
Comment (artcc) [reply frontend/src/lib/free-write-corrections.ts]: Thanks for the update and the additional tests. The allocation changes address the original repeated-fragment and `bin`/`in` examples, and the other three review points look addressed as well.

There is one remaining edge case in `findOccurrences()`: it returns as soon as the first matching variant has any hits, so an exact subword match prevents a case-insensitive whole-word match from ever becoming a candidate.

For example, with answer `Ich bin In Berlin` and `original: "in"`, the exact search finds `in` inside `bin` and returns immediately. The standalone `In` is never considered, so the whole-word priority in the allocation pass cannot select it. The same early return can leave repeated fragments with mixed capitalization unannotated.

Could you adjust candidate collection so eligible case-insensitive/trimmed occurrences remain available when needed, while preserving exact-match preference among otherwise equivalent candidates and deduplicating ranges? Please add a regression test combining the subword collision with mixed capitalization, plus repeated fragments with different capitalization.

The schema filtering, lesson-page tests, and revised amber/fallback behavior look good. I will still handle documentation and release bookkeeping as agreed. One small PR-description detail: the fallback screenshot caption still says amber, although the updated implementation now keeps that case red.

Thanks again for working through the feedback.

**Comment by artcc** [comment]
Thanks again for the work on this PR and for addressing the review feedback.

We cannot merge it yet because `develop` has moved forward with changes that are already in production, and those changes now conflict with this branch in `frontend/src/app/(app)/lesson/[id]/page.tsx`. In particular, the evaluated-answer block now uses icon components, while this PR changes the same block to add inline annotations and the partial-answer state.

Could you please bring the latest `develop` into your branch and resolve the conflict, preserving both the current production UI changes and the new corrections functionality? The lesson-page tests may also need adjusting, since some assertions currently depend on the literal status symbols rather than the icon-based rendering. Please rerun the relevant checks after integrating the changes.

The follow-up matching issue from our September 7 comment also remains pending in the currently published head (`8e509d2`): an exact subword match can prevent a case-insensitive whole-word occurrence from being considered, for example `original: "in"` in `Ich bin In Berlin`. Please include that adjustment and the regression tests requested in the thread.

Once those updates are in place and CI is green, we can review the integrated result. I will continue to handle the documentation and release bookkeeping as agreed. Thank you!

**Comment by arqo123** [comment]
Merged the latest `develop` into this branch — the conflict is resolved and the PR is mergeable again.

The only conflict was in `frontend/src/app/(app)/lesson/[id]/page.tsx`, where the new palette and accessible status icons met the corrections display:

- The answer field and the annotated answer block now use the `fl-success` / `fl-error-fg` tokens instead of the raw green/red Tailwind colors, and the inline `del` / `ins` markers follow the same tokens.
- The score badge keeps the new `role="img"` + `aria-label` treatment with the `Check` / `X` icons, plus a `Diff` icon for the partially correct state, labelled with the existing `corrections` key.
- The partial state still uses `amber-500` / `amber-400`, since there is no `fl-*` warning token yet and amber remains in use elsewhere in the app. Happy to switch it to a new token if you'd rather introduce one.

The corrections tests were updated to assert on the labelled badge and the new token classes instead of the old glyphs.

Verified locally on the merge commit: `./scripts/format.sh`, `npx tsc --noEmit`, `npm run lint`, `npm run test:run` (511 passing) and `pytest` (1023 passing, 85.58% coverage).
